ksqldb-engine/src/main/java/io/confluent/ksql/query/QueryExecutor.java

```diff
@@ -261,30 +274,69 @@ PersistentQueryMetadata buildPersistentQuery(
         .map(userErrorClassifiers::and)
         .orElse(userErrorClassifiers);
 
-    return new PersistentQueryMetadataImpl(
-        statementText,
-        querySchema,
-        sources,
-        sinkDataSource,
-        planSummary,
-        queryId,
-        materializationProviderBuilder,
-        applicationId,
-        topology,
+    final boolean useBinPacked = true;
+    if (useBinPacked) {
+      SharedKafkaStreamsRuntime sharedKafkaStreamsRuntime = getStream(sources);
+      PersistentQueriesInSharedRuntimesImpl binPackedPersistentQueryMetadata = new PersistentQueriesInSharedRuntimesImpl(
+          statementText,
+          querySchema,
+          sources,
+          planSummary,
+          applicationId,
+          topology,
+              sharedKafkaStreamsRuntime,
+          runtimeBuildContext.getSchemas(),
+          config.getOverrides(),
+          queryId,
+          materializationProviderBuilder,
+          physicalPlan,
+          getUncaughtExceptionProcessingLogger(queryId),
+          sinkDataSource,
+          listener
+      );
+      sharedKafkaStreamsRuntime.addQuery(classifier, streamsProperties, binPackedPersistentQueryMetadata, queryId);
+      return binPackedPersistentQueryMetadata;
+    } else {
+
+      return new PersistentQueryMetadataImpl(
+          statementText,
+          querySchema,
+          sources,
+          sinkDataSource,
+          planSummary,
+          queryId,
+          materializationProviderBuilder,
+          applicationId,
+          topology,
+          kafkaStreamsBuilder,
+          runtimeBuildContext.getSchemas(),
+          streamsProperties,
+          config.getOverrides(),
+          ksqlConfig.getLong(KSQL_SHUTDOWN_TIMEOUT_MS_CONFIG),
+          classifier,
+          physicalPlan,
+          ksqlConfig.getInt(KsqlConfig.KSQL_QUERY_ERROR_MAX_QUEUE_SIZE),
+          getUncaughtExceptionProcessingLogger(queryId),
+          ksqlConfig.getLong(KsqlConfig.KSQL_QUERY_RETRY_BACKOFF_INITIAL_MS),
+          ksqlConfig.getLong(KsqlConfig.KSQL_QUERY_RETRY_BACKOFF_MAX_MS),
+          listener,
+          scalablePushRegistry
+      );
+    }
+  }
+
+  private SharedKafkaStreamsRuntime getStream(final Set<SourceName> sources) {
```

Comment:
@rodesai I agree that we should move this. I don't think we need to do that right now though. I would like to keep it as simple as possible and I think once we get it all in one runtime we will not even need this. If there is an intermediate step I propose we move it then